src/vm/arm/PInvokeStubs.asm

```diff
@@ -107,6 +114,102 @@ __PInvokeGenStubFuncName SETS "$__PInvokeGenStubFuncName":CC:"_RetBuffArg"
 
     TEXTAREA
 ; ------------------------------------------------------------------
+; JIT_PInvokeBegin helper
+;
+; in:
+; r0 = InlinedCallFrame*: pointer to the InlinedCallFrame data, including the GS cookie slot (GS cookie right 
+;                         before actual InlinedCallFrame data)
+; 
+        NESTED_ENTRY JIT_PInvokeBegin
+
+            PROLOG_PUSH         {r4,r5,r7,lr}
+            PROLOG_STACK_SAVE   r7
+            
+            ldr     r1, =s_gsCookie
+            ldr     r1, [r1]
+            str     r1, [r0]
+            add     r4, r0, SIZEOF__GSCookie
+                        
+            ;; r4 = pFrame
+            
+            ;; set first slot to the value of InlinedCallFrame::`vftable' (checked by runtime code)
+            ldr     r1, =$InlinedCallFrame_vftable
+            str     r1, [r4]
+
+            mov     r1, 0
+            str     r1, [r4, #InlinedCallFrame__m_Datum]
+        
+            add     r1, sp, 0x10
+            str     r1, [r4, #InlinedCallFrame__m_pCallSiteSP]
+            str     r11, [r4, #InlinedCallFrame__m_pCalleeSavedFP]
+            str     lr, [r4, #InlinedCallFrame__m_pCallerReturnAddress]
+
+            bl      GetThread
```

Comment:
You should use `INLINE_GETTHREAD` and similar macros to fetch the thread inline. The call is too expensive.

Reply:
There's no INLINE_GETTHREAD for arm/arm64. I'll look into adding one

Reply:
You can copy the core of it from CoreRT.